### ISSUE TITLE
shellcheck fails recognize printf '%s\n' when eval'd.

### ISSUE THREAD

**Issue by orbea** [koalaman/shellcheck#1000]
#### For bugs
- Rule Id: `SC1117`
- My shellcheck version:  `online`
- [x] I tried on shellcheck.net and verified that this is still a problem on the latest commit
- [ ] It's not reproducible on shellcheck.net, but I think that's because it's an OS, configuration or encoding issue

#### Here's a snippet or screenshot that shows the problem:

```sh

#!/bin/sh

set -eu

foo=d

set -- a b c '$foo'

printf '%s\n' "$4" # $foo

eval "printf '%s\n' \"$4\"" # d

```

#### Here's what shellcheck currently says:

```
Line 11:
eval "printf '%s\n' \"$4\""
                ^-- SC1117: Backslash is literal in "\n". Prefer explicit escaping: "\\n".
```

#### Here's what I wanted or expected to see:

Nothing.

The advice for the rule is to use `"%s\\n"` or '%s\n' and the example is using the latter. However shellcheck does not realize that because the variable is being expanded with `eval` and that requires wrapping the entire `printf` command with double quotes.

**Comment by koalaman** [comment]
ShellCheck is right. The single quotes here are inside double quotes, and they therefore don't affect quoting or escaping. The `\n` is in a double quoted context. This happens with or without `eval`. 

Just like `"`s, the `\` should be escaped when being embedded as a literal in a double quoted string.

**Comment by orbea** [comment]
I stand corrected, thanks for the explanation!

**Comment by Byron-TW** [comment]
I ran into something similar, and realized that what I wanted is this:
```sh
echo $'hello\nworld'
```
which produces 
```
hello
world
```

**Comment by orbea** [comment]
@Byron-TW I suggest using printf instead, this is much more portable. For example.
```
printf %s\\n 'hello' 'world'
```
